src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,363 @@
+            private readonly SyntaxNode _localStatement;
+            private readonly SyntaxNode _enclosingBlock;
+            private readonly CancellationToken _cancellationToken;
+
+            public Analyzer(
+                SemanticModel semanticModel,
+                ILocalSymbol localSymbol,
+                SyntaxNode comparison,
+                SyntaxNode operand,
+                SyntaxNode localStatement,
+                SyntaxNode enclosingBlock,
+                CancellationToken cancellationToken)
+            {
+                Debug.Assert(semanticModel != null);
+                Debug.Assert(localSymbol != null);
+                Debug.Assert(comparison != null);
+                Debug.Assert(operand != null);
+                Debug.Assert(localStatement != null);
+                Debug.Assert(enclosingBlock != null);
+
+                _semanticModel = semanticModel;
+                _comparison = comparison;
+                _localSymbol = localSymbol;
+                _operand = operand;
+                _localStatement = localStatement;
+                _enclosingBlock = enclosingBlock;
+                _cancellationToken = cancellationToken;
+            }
+
+            // To convert a null-check to pattern-matching, we should make sure of a few things:
+            //
+            //      (1) The pattern variable may not be used before the point of declaration.
+            //
+            //          {
+            //              var use = t;
+            //              if (x is T t) {}
+            //          }
+            //
+            //      (2) The pattern variable may not be used outside of the new scope which
+            //          is determined by the parent statement.
+            //
+            //          {
+            //              if (x is T t) {}
+            //          }
+            //
+            //          var use = t;
+            //
+            //      (3) The pattern variable may not be used before assignment in opposite
+            //          branches, if any.
+            //
+            //          {
+            //              if (x is T t) {}
+            //              var use = t;
+            //          }
+            //
+            // We walk up the tree from the point of null-check and see if any of the above is violated.
+            public bool CanSafelyConvertToPatternMatching()
+            {
+                // Keep track of whether the pattern variable is definitely assigned when false/true.
+                // We start by the null-check itself, if it's compared with '==', the pattern variable
+                // will be definitely assigned when false, because we wrap the is-operator in a !-operator.
+                var defAssignedWhenTrue = _comparison.Kind() == SyntaxKind.NotEqualsExpression;
+
+                foreach (var current in _comparison.Ancestors())
+                {
+                    switch (current.Kind())
```

Comment:
>switch [](start = 20, length = 6)

A comment explaining what category of syntax should be handled here would be useful. When we add new syntax, it'll be useful to have some idea whether a new case should be added here.

I second @CyrusNajmabadi's concern with implementing something like flow analysis, but I don't have a much better idea...